content/post/walkthrough-exploiting-bugs-to-gain-code-execution.md

```diff
@@ -0,0 +1,327 @@
++++
+author = "cbrown"
+date = "2015-09-18T19:55:54-07:00"
+description = ""
+tags = ["exploit", "rop", "aslr", "shell"]
+title = "Walkthrough: Exploiting bugs to gain code execution on a modern operating system"
+
++++
+
+Whenever I discover a bug or typo in code I've written, I like to ask myself what the worst case effects of that typo might be. The most interesting question of course, is "If someone else found this before I did, could they have used it to gain arbitrary code execution?"
+
+We've seen arbitrary *root* code execution gained from as little as a [single random bit being erroneously flipped](https://www.blackhat.com/docs/us-15/materials/us-15-Seaborn-Exploiting-The-DRAM-Rowhammer-Bug-To-Gain-Kernel-Privileges-wp.pdf), so in cases of buffer overflow or use-after-free errors, I tend to think the answer is virtually always yes.
+
+But as an exercise, I occasionally feel a need to actually develop exploits, and if you're into C++, it's a great learning experience. Due to advances in binary security there are an increasing number of hurdles, but I'm gonna walk you through the steps to exploit a simple overflow on a modern operating system.
+
+**Prerequisites:** Basic knowledge of how a [call stack](https://en.wikipedia.org/wiki/Call_stack) works, basic understanding of common assembly instructions (push, pop, mov, add...). Being comfortable debugging executables is a plus.
+
+My payload will be for OSX Yosemite, but the same techniques apply to other systems.
+
+The vulnerability
+-----------------
+
+First, we'll need a vulnerable program. I'm gonna write up a quick one for this, but if you want to go hunting for these kinds of things in the wild, by far the easiest way is to grab a copy of [AFL](http://lcamtuf.coredump.cx/afl/) and start fuzzing open source projects.
+
+Our vulnerable program: https://gist.github.com/ccbrown/ff9925a58c2e45da2b5b
+
+The important part is here:
+
+```
+int run() {
+    while (true) {
+        ssize_t bytes = recv(_fd, _recvBuffer + _recvBytes, sizeof(_recvBuffer), 0);
+        if (bytes <= 0) {
+            break;
+        }
+        _recvBytes += bytes;
+        void* newline = memchr(_recvBuffer, '\n', _recvBytes);
+        if (newline) {
+            send(_fd, _recvBuffer, (char*)newline - _recvBuffer + 1, 0);
+            break;
+        }
+    }
+    return 0;
+}
+```
```

Comment:
Yeah I was a lot more concerned about minimizing code block size before I reduced the font.